### PR TITLE
Fix new dynlink's initialisation checks

### PR DESCRIPTION
**Pull request by lpw25**
#1063 rewrote the various checks when loading shared libraries with dynlink in order to make them sound. However, the checks to ensure that a module is not used before it has been initialised were far too restrictive. For example, a module in a shared library could not refer to another library within the same shared library. This PR fixes that by updating the list of initialised units earlier. This PR also relaxes the "allowed units" check so that it only applies to implementation dependencies not interface dependencies -- which matches the previous behaviour and is necessary to deal with mli-only modules.

I also added a bunch of tests that exercise the various cases.

This PR should hopefully fix MPR#7876 -- although I haven't actually tried it.

This PR is on top of #2175 because that bug was causing problems with the use of packing added by #1063.

### CODE REVIEW

**Comment by ejgallego** [comment]
I confirm this fixes Coq plugins in both build modes [using -pack and using aliases].

**Comment by gasche** [comment]
cc @alainfrisch @chambart @mshinwell : we don't want to wait for too long between the merge of #1063 in trunk and a fix such as included in the present PR, so I think the urgency of this review is a bit higher than the usual.

**Comment by mshinwell** [comment]
Ack, I will read this soon.

**Comment by gasche** [comment]
( @lpw25, you could rebase now that #2175 is merged. )

**Comment by lpw25** [comment]
Rebased

File: testsuite/tests/lib-dynlink-initializers/test2_main.ml
Comment (mshinwell) [inline testsuite/tests/lib-dynlink-initializers/test2_main.ml]: Please clarify what "its" refers to.  (And, "its" -> "it's".)

File: testsuite/tests/lib-dynlink-initializers/test3_main.ml
Comment (mshinwell) [inline testsuite/tests/lib-dynlink-initializers/test3_main.ml]: Same comment as above regarding "it is".

File: testsuite/tests/lib-dynlink-initializers/test5_main.ml
Comment (mshinwell) [inline testsuite/tests/lib-dynlink-initializers/test5_main.ml]: Typo "libary".  Also maybe name the libraries "A" and "B" so we can make "that library" precise.

**Review by mshinwell — CHANGES_REQUESTED** [review]

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
This is OK to merge after the `Changes` file has been updated.

**Pull request merged**
